### PR TITLE
munger: Add a on option for e2e status checker to read data from GCS instead of Jenkins

### PR DESCRIPTION
**Pull request by gmarek**
It's first of the hacks I need to get it to add more complex logic for handling scalability test failures. It's no-op as it currently does a separate check and only prints errors if it finds differences between Jenkins and GCS. We should keep it running for few days before switching to GCS only.

It's a small hack that I need, as @justinsb PRs are stalled. It's small though, so it should be pretty straightforward to review - @eparis can you review the changes from the second commit only? It'd be great if you could do this sooner rather than later.

cc @kubernetes/sig-scalability @kubernetes/sig-testing @spxtr @wojtek-t

### CODE REVIEW

File: mungegithub/mungers/e2e/e2e.go [hunk 1]
```diff
@@ -70,21 +77,142 @@ func (e *E2ETester) Stable() bool {
...
+			path := fmt.Sprintf("artifacts/junit_%02d.xml", i)
+			response, err := utils.GetFileFromJenkinsGoogleBucket(job, lastBuildNumber, path)
+			if err != nil {
+				glog.Errorf("Error while getting data for %v/%v/%v: %v", job, lastBuildNumber, path, err)
+				continue
+			}
+			if response.StatusCode != 200 {
+				break
+			}
+			{
```
Comment (eparis) [inline mungegithub/mungers/e2e/e2e.go]: So what's this about? It looks like you are trying to get the defered Close() to run, but that's not how golang works....

http://play.golang.org/p/eU4sDqT42P

File: mungegithub/mungegithub.go [hunk 2]
```diff
@@ -103,6 +106,33 @@ func main() {
 			if err := config.Features.Initialize(requestedFeatures); err != nil {
 				return err
 			}
+			if config.E2EStatusCheckOnly {
```
Comment (eparis) [inline mungegithub/mungegithub.go]: I don't like this whole block. I'd much rather testing be done with `--pr-mungers=submit-queue --min-pr-number=0 --max-pr-number=0 --once`.  And get this output via a `--v=N` flag....

This is really uglying up a core function with very specific test only code...

File: mungegithub/mungers/submit-queue.go [hunk 3]
```diff
@@ -230,6 +235,11 @@ func (sq *SubmitQueue) e2eStable() bool {
 	wentUnstable := false
 
 	stable := sq.e2e.Stable()
+	gcsStable := sq.e2e.GCSBasedStable()
+
+	if !gcsStable {
+		glog.Errorf("Weak stable check returned error.")
```
Comment (eparis) [inline mungegithub/mungers/submit-queue.go]: Is this an error? seems like stable != gcsStable is an error. But gcsStable == false seem 'normal'....

File: mungegithub/mungers/e2e/e2e.go [hunk 4]
```diff
@@ -70,21 +77,142 @@ func (e *E2ETester) Stable() bool {
...
+	}
+	glog.Infof("%v", result)
+	return result.Result == successString
+}
+
+// GCSBasedStable is a version of Stable function that depends on files stored in GCS instead of Jenkis
+func (e *E2ETester) GCSBasedStable() bool {
+	for _, job := range e.JobNames {
+		lastBuildNumber, err := utils.GetLastestBuildNumberFromJenkinsGoogleBucket(job)
+		glog.Infof("Checking status of %v, %v", job, lastBuildNumber)
```
Comment (eparis) [inline mungegithub/mungers/e2e/e2e.go]: Way too high log level

File: mungegithub/mungers/e2e/e2e.go [hunk 5]
```diff
@@ -70,21 +77,142 @@ func (e *E2ETester) Stable() bool {
...
+
+func checkFinishedStatus(job string, lastBuildNumber int) bool {
+	response, err := utils.GetFileFromJenkinsGoogleBucket(job, lastBuildNumber, "finished.json")
+	if err != nil {
+		glog.Errorf("Error while getting data for %v/%v/%v: %v", job, lastBuildNumber, "finished.json", err)
+		return false
+	}
+
+	defer response.Body.Close()
+	if response.StatusCode != 200 {
```
Comment (eparis) [inline mungegithub/mungers/e2e/e2e.go]: http.StatusOK

File: mungegithub/mungers/e2e/e2e.go [hunk 6]
```diff
@@ -70,21 +77,142 @@ func (e *E2ETester) Stable() bool {
...
+	}
+	result := finishedFile{}
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.Errorf("Failed to read the response for %v/%v/%v: %v", job, lastBuildNumber, "finished.json", err)
+		return false
+	}
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		glog.Errorf("Failed to unmarshal %v: %v", string(body), err)
```
Comment (eparis) [inline mungegithub/mungers/e2e/e2e.go]: Should we return here?

File: mungegithub/mungers/e2e/e2e.go [hunk 7]
```diff
@@ -70,21 +77,142 @@ func (e *E2ETester) Stable() bool {
...
+				body, err = reader.ReadString('\n')
+				if err != nil {
+					glog.Errorf("Failed to read the response for %v/%v/%v: %v", job, lastBuildNumber, path, err)
+					continue
+				}
+				numberOfTests := 0
+				nubmerOfFailures := 0
+				timestamp := 0.0
+				fmt.Sscanf(strings.TrimSpace(body), "<testsuite tests=\"%d\" failures=\"%d\" time=\"%f\">", &numberOfTests, &nubmerOfFailures, &timestamp)
+				glog.Infof("%v, numberOfTests: %v, numberOfFailures: %v", string(body), numberOfTests, nubmerOfFailures)
```
Comment (eparis) [inline mungegithub/mungers/e2e/e2e.go]: Log Level.

**Comment by eparis** [comment]
Is GCSWeakStable something we'd want all tests to move to? Should infrastructure failures on other tests really cause the queue to block?

**Comment by gmarek** [comment]
I'll address your comments on Monday. We certainly want to move long running scalability tests. I'm not sure about other ones.

File: mungegithub/mungers/e2e/e2e.go (hunk 1 above)
Comment (gmarek) [inline mungegithub/mungers/e2e/e2e.go]: Yeah... right. IIUC go have an early binding, so it should be OK with plain `defer` here... I hope.

File: mungegithub/mungegithub.go (hunk 2 above)
Comment (gmarek) [inline mungegithub/mungegithub.go]: OK, I'll remove this, but it's currently so very ugly coupled with everything else...

File: mungegithub/mungers/submit-queue.go (hunk 3 above)
Comment (gmarek) [inline mungegithub/mungers/submit-queue.go]: Yes, sorry - it's a debugging leftover.

File: mungegithub/mungers/e2e/e2e.go (hunk 4 above)
Comment (gmarek) [inline mungegithub/mungers/e2e/e2e.go]: Done.

File: mungegithub/mungers/e2e/e2e.go (hunk 5 above)
Comment (gmarek) [inline mungegithub/mungers/e2e/e2e.go]: Done.

File: mungegithub/mungers/e2e/e2e.go (hunk 6 above)
Comment (gmarek) [inline mungegithub/mungers/e2e/e2e.go]: Yes.

File: mungegithub/mungers/e2e/e2e.go (hunk 7 above)
Comment (gmarek) [inline mungegithub/mungers/e2e/e2e.go]: Done.

**Comment by gmarek** [comment]
@eparis PTAL

**Comment by gmarek** [comment]
@k8s-bot unit test this issue: #IGNORE
`remote: Daily bandwidth rate limit exceeded for IP 52.22.60.255.`

File: mungegithub/features/features.go [hunk 8]
```diff
@@ -47,7 +47,8 @@ func (f *Features) GetActive() []feature {
 // Initialize should be called with the set of all features needed by all (active) mungers
 func (f *Features) Initialize(requestedFeatures []string) error {
 	for _, name := range requestedFeatures {
-		feat, found := featureMap[name]
+		glog.Infof("Initilizing feature: %v", name)
+    feat, found := featureMap[name]
```
Comment (eparis) [inline mungegithub/features/features.go]: This seem gofmt questionable

File: mungegithub/mungegithub.go [hunk 9]
```diff
@@ -38,11 +38,11 @@ var (
...
-	MinIssueNumber   int
-	PRMungersList    []string
-	IssueReportsList []string
-	Once             bool
-	Period           time.Duration
+	MinIssueNumber     int
+	PRMungersList      []string
+	IssueReportsList   []string
+	Once               bool
+	Period             time.Duration
```
Comment (eparis) [inline mungegithub/mungegithub.go]: Not sure about these either...

**Comment by gmarek** [comment]
Fascinating - I'm completely sure I run gofmt before pushing changes...

**Comment by gmarek** [comment]
Travis seems to be permanently unhappy today. @eparis PTAL

**Comment by eparis** [comment]
True, travis is unhappy. But the one that ran showed that you didn't run `godep go test ./...`
```
mungers/submit-queue_test.go:154: sq.JenkinsJobs undefined (type *SubmitQueue has no field or method JenkinsJobs)
# k8s.io/contrib/mungegithub/mungers/e2e
```

**Comment by gmarek** [comment]
Sorry for that. @eparis PTAL.

**Comment by eparis** [comment]
@justinsb I'm merging this. Your federated builder PR is going to need a bit larger rebase This also pulls from GCS so we are going to need to reconcile the two. I really really like the federated builder PR.

**Pull request merged**